Add basler_synced_cameras_node for running multiple synced cameras

petercheng00 opened the pull request:
This is the specialized and much simpler/cleaner approach to syncing I mentioned. A lot of the changes are just extracting out the basler parameter-handling code to its own file.
Config files are split between a "cameras" section, which is a list of dictionaries, and then the existing basler parameters, which are identical for each camera.
By ensuring that cameras are all synced at the same time to the same master, we should avoid race condition problems.

petercheng00 committed:
Add basler_synced_cameras_node for running multiple synced cameras

petercheng00 committed:
Add option for image_publisher to use camera's timestamp

bhomberg commented on src/basler_parameters.cpp:
```diff
@@ -0,0 +1,151 @@
+            return;
+        }
+        this_node->FromString(value.c_str());
+        //ROS_INFO_STREAM(name << " set to " << this_node->GetValue());
```
Delete comment or uncomment and switch to DEBUG

bhomberg commented on src/basler_parameters.cpp:
```diff
@@ -0,0 +1,151 @@
+            return;
+        }
+        this_node->SetValue(value);
+        ROS_INFO_STREAM(name << " set to " << this_node->GetValue());
```
I know you were talking about log levels earlier -- do you think these need to stay INFO's or should they be DEBUG's?

bhomberg commented on include/basler_parameters.h:
```diff
@@ -0,0 +1,3 @@
+#include <pylon/PylonIncludes.h>
+
+void handle_basler_parameters(Pylon::CInstantCamera& camera);
```
If the other functions aren't going to be exposed, it might be helpful to clarify that somehow (e.g., add a detail:: namespace).

bhomberg commented on include/image_publisher.h:
```diff
@@ -11,6 +11,8 @@
 
 #include <ros/console.h>
```
Since you're touching the includes, can you sort + alphebetize within sets? 

https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

bhomberg commented on src/basler_synced_cameras_node.cpp:
```diff
@@ -0,0 +1,255 @@
+
+void enableSyncFreeRun(Pylon::CBaslerGigEInstantCamera& camera, float frame_rate)
+{
+    // See https://docs.baslerweb.com/synchronous-free-run.html for info.
```
Can you put the tl;dr in the comment for readers who just need the high level info?

bhomberg commented on src/basler_synced_cameras_node.cpp:
```diff
@@ -0,0 +1,255 @@
+    {
+        camera.Open();
+        handle_basler_parameters(camera);
+        camera.GevIEEE1588.SetValue(true);
```
comment what this is?

bhomberg commented on src/basler_synced_cameras_node.cpp:
```diff
@@ -0,0 +1,255 @@
+            }
+            if (!camera_found)
+            {
+                ROS_INFO_STREAM("Failed to find camera " << index << " with serial " << serial);
```
It looks like you're looping here, so this is just one instance of a missing camera and if a camera is just unplugged or someone mistyped the serial, this code will loop forever.  This would be more usable with a timeout + failure/shut down with an actual error after the timeout has passed.  

Also, I'd argue that it's not helpful to print "failed to find camera" on every loop -- if it takes a few seconds sometimes to find a camera, the user doesn't really need to know that, if it's expected -- they just want to see when the camera is eventually found or what the error was after it times out.  Maybe switch this to debug?

bhomberg commented on src/basler_synced_cameras_node.cpp:
```diff
@@ -0,0 +1,255 @@
+{
+    ROS_INFO("Waiting for %zu cameras to be PTP slaves...", cameras.size());
+    ros::Rate r(0.5);
+    while (ros::ok())
```
You should have a timeout for this, too.  The default timeout can be pretty long (I know you said it might take a minute or two to get set up) but at some point, it should fail.  Ideally the overall timeout would be user-configurable.

bhomberg commented on src/basler_synced_cameras_node.cpp:
```diff
@@ -0,0 +1,255 @@
+            case Basler_GigECamera::GevIEEE1588StatusLatched_Slave:
+            {
+                int64_t master_id = camera.GevIEEE1588ParentClockId();
+                if (num_slave == 0)
```
Can you add a comment explaining the logic here?  I think I worked it out correctly, but the code would be easier to read with an explanation.

bhomberg commented on src/basler_synced_cameras_node.cpp:
```diff
@@ -0,0 +1,255 @@
+        }
+        else
+        {
+            ROS_INFO("Camera PTP status: %d initializing, %d masters, %d slaves", num_init, num_master, num_slave);
```
Can you throttle these outputs?  Maybe max one every 10s?  Otherwise if it takes a while to sync, that'll be a lot of messages!

You can also output it just the first time (or switch to debug).

options: http://wiki.ros.org/roscpp/Overview/Logging

bhomberg commented on src/basler_synced_cameras_node.cpp:
```diff
@@ -0,0 +1,255 @@
+    }
+}
+
+void waitForPTPClockSync(const std::vector<Pylon::CBaslerGigEInstantCamera>& cameras, int max_offset_ns, int offset_window_s)
```
Can you add a comment explaining what this function does.

bhomberg commented on src/basler_synced_cameras_node.cpp:
```diff
@@ -0,0 +1,255 @@
+
+void waitForPTPClockSync(const std::vector<Pylon::CBaslerGigEInstantCamera>& cameras, int max_offset_ns, int offset_window_s)
+{
+    ROS_INFO("Waiting for clock offsets < %d ns over %d s...", max_offset_ns, offset_window_s);
```
Can you put this info in the message at the end instead and switch this to debug?

bhomberg commented on src/basler_synced_cameras_node.cpp:
```diff
@@ -0,0 +1,255 @@
+                currently_below = true;
+                below_start = ros::Time::now();
+            }
+            else if ((ros::Time::now() - below_start).toSec() >= offset_window_s)
```
Can you comment to clarify what's happening in each of these cases / what the logic is?

bhomberg commented on src/basler_synced_cameras_node.cpp:
```diff
@@ -0,0 +1,255 @@
+            currently_below = false;
+        }
+        // Print out current offsets periodically
+        if ((ros::Time::now()  - last_output).toSec() >= 1)
```
This should definitely be DEBUG level.  You may also be able to use THROTTLE to handle this nicely for you, at least if you can package it all up in one print statement.

bhomberg reviewed (changes requested):
This is great progress!  A few comments, mostly on 3 topics:

-More comments.  Some of this logic is tricky, and would be great to explain how it's working so the user can verify the lack of race conditions themselves.
-Add timeouts -- this code shouldn't hang forever if something is wrong.
-Decrease log level.  It would probably be fine for just a camera system, but as a part of our full robot systems with a lot of other spammy things (especially messages interrupting between camera-related outputs), it's going to be clutter. Users will care if it worked/didn't, and if not, why (for the most part); basically everything else can be DEBUG.

One more thought because of the tricky logic -- how hard would it be to add some tests?  (@jonbinney , do we even have a way to run tests on this library?)  It would be great to have some automated verification that the logic is behaving as you intend.

petercheng00 committed:
Consolidate code in basler_parameters.cpp

petercheng00 committed:
Reorder includes in image_publisher.h

petercheng00 committed:
Add comments and reduce log level in basler_synced_cameras_node

petercheng00 commented:
@bhomberg thanks for feedback, I think I've addressed the comments. I've added more comments, decreased logging, and added timeouts throughout. I also went through and cleaned up basler_parameters.cpp a bit.

bhomberg commented on src/basler_parameters.cpp:
```diff
@@ -0,0 +1,108 @@
+        }
+        else
+        {
+            ROS_FATAL_STREAM("Unknown param type for parameter " << name << ": " << type);
```
Why FATAL here instead of ERROR -- it looks like you use error above?  I feel like we normally use ERROR for this kind of thing.  I believe we only use FATAL right before killing the node (maybe that's the case here, but it'd be clearer if the FATAL statement is right before the node closes).

petercheng00 replied:
This is from the original code - so far I've kept the behavior of the code in basler_parameters identical. I can change it to be error (and return false) though, and handle the exiting in the nodes.

petercheng00 replied:
On 2nd thought - there's a lot of ROS_ASSERTs in this file (also all from pre-existing code), to catch unexpected yaml settings. This is also catching a bad yaml entry, so in that light, having a FATAL and crashing here seems more in line with the rest of the code - thoughts?

jonbinney replied:
+1 for crashing on bad inputs. I've forgotten - does ROS_FATAL also kill the node?

bhomberg replied:
I was looking and it doesn't seem like it does. I'm definitely not opposed to crashing the node on bad input -- just think it makes sense to do it in the same place as you announce the fatal error.  (If I misread documentation, though, that solves that problem).

petercheng00 replied:
Just checked, and ROS_FATAL does kill the node (node no longer shows up in rosnode list). ROS_ASSERT however does not kill the node.

bhomberg commented on src/basler_synced_cameras_node.cpp:
```diff
@@ -0,0 +1,293 @@
+                {
+                    ROS_DEBUG_STREAM(devices[i].GetSerialNumber());
+                }
+                if (ros::Time::now() > timeout_time)
```
Do you think it makes more sense as an individual timeout for each camera or an overall timeout for connecting to all of the cameras?  Might be slightly cleaner, but this would obviously time out faster if you have a large number of sync'd cameras.  (Not a huge preference here, though, so if you prefer this that's fine.)

petercheng00 replied:
In practice I think the effect is largely the same. When a camera is present, the time to connect is pretty fast, so the timeout only comes into effect for the first missing camera. I haven't actually seen cases where it takes some time to connect, but if that's a thing that happens, I don't think the time one camera takes to connect should affect the time allotted for another camera.

bhomberg commented on src/basler_synced_cameras_node.cpp:
```diff
@@ -0,0 +1,293 @@
+{
+    ROS_DEBUG("Waiting for %zu cameras to be PTP slaves...", cameras.size());
+    ros::Rate r(0.5);
+    ros::Time timeout_time = ros::Time::now() + ros::Duration(60);
```
Could you set these default timeouts via a constant variable at the top of the file (or from a config file if there is one)?  I'm not sure they need to be passed in through the launch file (though that might be helpful if you think users might want to change the timeouts).

bhomberg reviewed (approved):
A couple minor comments but overall LGTM!
